File: DataBaseWithBusinessLogicConnector/Entities/OperationsGroup.cs
using DataBaseWithBusinessLogicConnector.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using DataBaseWithBusinessLogicConnector.Properties;

namespace DataBaseWithBusinessLogicConnector.Entities
{
    public class OperationsGroup : IEntity, IOperation
    {
        public int? Id { get; private set; }
        public bool IsDirty { get; set; }
        public bool IsMarkForDeletion { get; set; }
        public User User { get; private set; }
        public decimal Amount { get; private set; }
        public string Description { get; private set; }
        public Frequency Frequency { get; private set; }
        public Importance Importance { get; private set; }
        public TransactionType TransactionType { get; private set; }
        public DateTime Date { get; private set; }
        public List<RelTag> Tags { get; private set; }
        public List<Operation> Operations { get; private set; }

        public OperationsGroup(int? id, User user, string description, Frequency frequency, Importance importance, DateTime date, List<RelTag> tags, List<Operation> operations)
        {
            Id = id;
            User = user;
            Description = description;
            Frequency = frequency;
            Importance = importance;
            Date = date;
            Tags = tags;
            Operations = operations;
        }

        public void SetOperations(IEnumerable<Operation> operations)
        {
            Operations = operations.ToList();
            IsDirty = true;
        }

        public void AddOperation(Operation operation)
        {
            Operations.Add(operation);
        }

        public void RemoveOperation(Operation operation)
        {
            Operations.Remove(operation);
        }

        public void RemoveAllOperations()
        {
            Operations.Clear();
        }

        public void SetTags(IEnumerable<RelTag> tags)
        {
            Tags = tags.ToList();
            IsDirty = true;
        }

        public void AddTag(RelTag tag)
        {
            if (Tags.All(t => t.Tag.Text != tag.Tag.Text))
            {
                tag.IsDirty = true;
                Tags.Add(tag);
            }
            else
            {
                var oldTag = Tags.Where(t => t.Tag.Text == tag.Tag.Text).First();
                oldTag.IsDirty = false;
                oldTag.IsMarkForDeletion = false;
            }
        }

        public void MarkTagForDeletion(RelTag tag)
        {
            var existTag = Tags.Where(t => t.Tag.Text == tag.Tag.Text).FirstOrDefault();
            existTag.IsDirty = true;
            existTag.IsMarkForDeletion = true;
        }

        public void MarkAllTagsForDeletion()
        {
            for (int i = 0; i < Tags.Count; i++)
            {
                RelTag tag = Tags[i];
                tag.IsDirty = true;
                tag.IsMarkForDeletion = true;
            }
        }

        public void SetFrequence(Frequency frequency)
        {
            Frequency = frequency;
            IsDirty = true;
        }

        public void SetImportance(Importance importance)
        {
            Importance = importance;
            IsDirty = true;
        }

        public void UpdateId(int? id)
        {
            Id = id;
        }

        public override string ToString()
        {
            return Date.ToString("dd.MM.yyyy") + " " + Description;
        }

        public void SetDescription(string text)
        {
            Description = text;
            IsDirty = true;
        }

        public void SetDate(DateTime date)
        {
            Date = date;
            IsDirty = true;
        }

        public void UpdateAmount(List<TransactionType> types)
        {
            Amount = decimal.Zero;
            foreach(var item in Operations)
            {
                if (item.TransactionType.Text == Properties.strings.income)
                {
                    Amount += item.Amount;
                } else
                {
                    Amount -= item.Amount;
                }
            }
            if (Amount < 0)
            {
                TransactionType = types.First(t => t.Text == strings.expense);
                Amount = Math.Abs(Amount);
            }
            else
            {
                TransactionType = types.First(t => t.Text == strings.income);
            }
        }
    }
}

File: PayglWeb/ClientApp/app/entities/OperationsGroup.ts
import { Operation } from './Operation';
import { User, Frequency, Importance, TagRelation, Tag, TransactionType, TransferType } from './entities';
import { OperationLike } from './OperationLike';
import { Countable } from './Countable';

export class OperationsGroup implements OperationLike, Countable{
    Id: number;
    IsDirty: boolean;
    IsMarkForDeletion: boolean;
    User: User;
    Description: string;
    Frequency: Frequency;
    Importance: Importance;
    Date: string;
    Tags: TagRelation[];
    Operations: Operation[];

    Amount: number = 0;
    TransactionType: TransactionType;

    constructor(id?: number, user?: User, description?: string, frequency?: Frequency, importance?: Importance, date?: string, tags?: TagRelation[], operations?: Operation[]) {
        this.Id = id
        this.User = user
        this.Description = description
        this.Frequency = frequency
        this.Importance = importance
        this.Date = OperationsGroup.convertDate(date);
        this.Tags = tags
        if (operations != null) {
            this.Operations = operations.sort((a, b) => (a.Date > b.Date) ? 1 : -1)
        }
    }

    recalculate(transactionTypes: TransactionType[]) {
        this.Amount = 0;
        for (let operation of this.Operations) {
            if (operation.TransactionType.Text == "wydatek") {
                this.Amount -= operation.Amount;
            } else {
                this.Amount += operation.Amount;
            }        
        }
        if (this.Amount > 0) {
            this.TransactionType = transactionTypes.filter(t => t.Text == "przychód")[0];
        } else {
            this.TransactionType = transactionTypes.filter(t => t.Text == "wydatek")[0];
        }
        this.Amount = Math.abs(this.Amount);
    }

    static createFromJson(data: any, frequencies: Frequency[], importances: Importance[], tags: Tag[], transactionTypes: TransactionType[], transferTypes: TransferType[]): OperationsGroup {
        let operationsGroup = new OperationsGroup()
        operationsGroup.Operations = [];
        for (let operation of data.Operations) {
            operationsGroup.Operations.push(Operation.createFromJson(operation, frequencies, importances, tags, transactionTypes, transferTypes));
        }
        operationsGroup.Id = data.Id;
        operationsGroup.User = User.createFromJson(data.User);
        operationsGroup.Frequency = frequencies.filter(t => t.Id === data.Frequency.Id)[0];
        operationsGroup.Importance = importances.filter(t => t.Id === data.Importance.Id)[0];
        operationsGroup.Date = OperationsGroup.convertDate(data.Date);
        operationsGroup.Tags = [];
        for (let tag of data.Tags) {
            operationsGroup.Tags.push(TagRelation.createFromJson(tag, tags));
        }
        operationsGroup.Description = data.Description;
        operationsGroup.Operations = operationsGroup.Operations.sort((a, b) => (a.Date > b.Date) ? 1 : -1)
        operationsGroup.recalculate(transactionTypes);

        return operationsGroup
    }

    private static convertDate(st: string): string {
        if (st == null) {
            return ""
        }
        var pattern = /(\d{2})\.(\d{2})\.(\d{4})/;
        return st.replace(pattern, '$3-$2-$1');
    }
}
